src/bm_sim/P4Objects.cpp

```diff
@@ -422,6 +422,57 @@ P4Objects::add_primitive_to_action(const Json::Value &cfg_primitive,
     if (type == "hexstr") {
       const auto value_hexstr = cfg_parameter["value"].asString();
       action_fn->parameter_push_back_const(Data(value_hexstr));
+    } else if (type == "parameters_vector") {
+      std::vector<ActionParam> params_vector;
```

Comment:
Is it possible to move the logic to build the parameter vector to actions.h / actions.cpp, so we don't have to expose the "internals" of the actions module (such as the `ActionParam` class)?

Something like:
```
struct ActionParamVectorBuilder {
    ActionParamVectorBuilder(ActionFn *action_fn)
        : action_fn(action_fn) { }

    void push_back_action_data(int action_data_offset);
    void push_back_const(...);
    void push_back_field(...);

    std::vector<ActionParam> get() const { return param_vector; }

    ActionFn *action_fn;  // needed for push_back_const
    std::vector<ActionParam> param_vector;
};
```

Please let me know if you see an issue with that approach.